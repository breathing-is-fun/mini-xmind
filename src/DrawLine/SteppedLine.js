import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Line from './Line';

export default class SteppedLine extends PureComponent {
  static propTypes = {
    x0: PropTypes.number.isRequired,
    y0: PropTypes.number.isRequired,
    x1: PropTypes.number.isRequired,
    y1: PropTypes.number.isRequired,
    orientation: PropTypes.oneOf(['h', 'v']),
    borderColor: PropTypes.string,
    borderStyle: PropTypes.string,
    borderWidth: PropTypes.number,
    className: PropTypes.string,
    zIndex: PropTypes.number,
  };

  render() {
    if (this.props.orientation === 'h') {
      return this.renderHorizontal();
    }
    return this.renderVertical();
  }

  renderVertical() {
    const { x0, y0, x1, y1 } = this.props;

    const dx = x1 - x0;
    if (dx === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const y2 = (y0 + y1) / 2;

    const xOffset = dx > 0 ? borderWidth : 0;
    const minX = Math.min(x0, x1) - xOffset;
    const maxX = Math.max(x0, x1);

    return (
      <div className="react-steppedlineto">
        <Line {...this.props} x0={x0} y0={y0} x1={x0} y1={y2} />
        <Line {...this.props} x0={x1} y0={y1} x1={x1} y1={y2} />
        <Line {...this.props} x0={minX} y0={y2} x1={maxX} y1={y2} />
      </div>
    );
  }

  renderHorizontal() {
    const { x0, y0, x1, y1 } = this.props;

    const dy = y1 - y0;
    if (dy === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const x2 = (x0 + x1) / 2;

    const yOffset = dy < 0 ? borderWidth : 0;
    const minY = Math.min(y0, y1) - yOffset;
    const maxY = Math.max(y0, y1);

    return (
      <div className="react-steppedlineto">
        <Line {...this.props} x0={x0} y0={y0} x1={x2} y1={y0} />
        <Line {...this.props} x0={x1} y0={y1} x1={x2} y1={y1} />
        <Line {...this.props} x0={x2} y0={minY} x1={x2} y1={maxY} />
      </div>
    );
  }
}
