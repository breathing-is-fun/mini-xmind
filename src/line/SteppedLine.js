import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Line from './core';
const defaultBorderWidth = 1;

export default class SteppedLine extends PureComponent {
  static propTypes = {
    orientation: PropTypes.oneOf(['h', 'v', 'horizonal', 'vertical']),
    className: PropTypes.string,
    style: PropTypes.object,
  };

  static defaultProps = {
    orientation: 'h',
    className: '',
    style: {},
  };

  renderVertical = () => {
    const { x0, y0, x1, y1 } = this.props;

    const dx = x1 - x0;
    if (dx === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const y2 = (y0 + y1) / 2;

    const xOffset = dx > 0 ? borderWidth : 0;
    const minX = Math.min(x0, x1) - xOffset;
    const maxX = Math.max(x0, x1);

    return (
      <div className="stepped-lineto">
        <Line {...this.props} x0={x0} y0={y0} x1={x0} y1={y2} />
        <Line {...this.props} x0={x1} y0={y1} x1={x1} y1={y2} />
        <Line {...this.props} x0={minX} y0={y2} x1={maxX} y1={y2} />
      </div>
    );
  };

  renderHorizontal = () => {
    const { x0, y0, x1, y1 } = this.props;

    const dy = y1 - y0;
    if (dy === 0) {
      return <Line {...this.props} />;
    }

    const borderWidth = this.props.borderWidth || defaultBorderWidth;
    const x2 = (x0 + x1) / 2;

    const yOffset = dy < 0 ? borderWidth : 0;
    const minY = Math.min(y0, y1) - yOffset;
    const maxY = Math.max(y0, y1);

    return (
      <div className="stepped-line">
        <Line {...this.props} x0={x0} y0={y0} x1={x2} y1={y0} />
        <Line {...this.props} x0={x1} y0={y1} x1={x2} y1={y1} />
        <Line {...this.props} x0={x2} y0={minY} x1={x2} y1={maxY} />
      </div>
    );
  };

  render() {
    const { orientation } = this.props;
    let result;

    switch (orientation) {
      case 'horizonal':
      case 'h':
        result = this.renderHorizontal();
        break;

      case 'vertical':
      case 'v':
        result = this.renderVertical();
        break;
    }

    return result;
  }
}
