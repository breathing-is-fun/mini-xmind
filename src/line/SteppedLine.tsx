import React from 'react';
import Line from './Line';
import {
  getLineCoordinatesForHorizonal,
  getLineCoordinatesForVertical,
} from '../utils/LineUtil';

export type OrientationProps = 'horizonal' | 'vertical';
export interface SteppedLineProps {
  className?: string;
  from: any;
  to: any;
  offset?: { x: number; y: number };
  orientation?: OrientationProps;
}

const SteppedLine = ({
  from,
  to,
  offset = { x: 0, y: 0 },
  orientation,
}: SteppedLineProps) => {
  const renderHorizonal = () => {
    const {
      x0,
      x1,
      y0,
      y1,
      centerX,
      firstLineVisible,
      secondLineVisible,
      thirdLineVisible,
    } = getLineCoordinatesForHorizonal(from, to, offset);

    return (
      <div className="stepped-line">
        {firstLineVisible && <Line x0={x0} y0={y0} x1={centerX} y1={y0} />}
        {secondLineVisible && (
          <Line x0={centerX} y0={y0} x1={centerX} y1={y1} />
        )}
        {thirdLineVisible && <Line x0={centerX} y0={y1} x1={x1} y1={y1} />}
      </div>
    );
  };

  const renderVertical = () => {
    const {
      x0,
      x1,
      y0,
      y1,
      centerY,
      firstLineVisible,
      secondLineVisible,
      thirdLineVisible,
    } = getLineCoordinatesForVertical(from, to, offset);

    return (
      <div className="stepped-line">
        {firstLineVisible && <Line x0={x0} y0={y0} x1={x0} y1={centerY} />}
        {secondLineVisible && (
          <Line x0={x0} y0={centerY} x1={x1} y1={centerY} />
        )}
        {thirdLineVisible && <Line x0={x1} y0={centerY} x1={x1} y1={y1} />}
      </div>
    );
  };

  if (orientation === 'horizonal') {
    return renderHorizonal();
  }

  if (orientation === 'vertical') {
    return renderVertical();
  }

  return renderHorizonal();
};

export default SteppedLine;
