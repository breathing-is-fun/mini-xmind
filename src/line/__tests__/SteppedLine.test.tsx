import React from 'react';
import { mount } from 'enzyme';
import SteppedLine from '../SteppedLine';
import 'nino-cli/scripts/setup';

describe('SteppedLine', () => {
  it('renderVertical should work well', () => {
    const wrapper = mount(
      <SteppedLine x0={1} y0={1} x1={0} y1={0} orientation="vertical" />,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.line-placeholder-item').length).toBe(3);
    const getTargetIndexProps = (wrapper, index) =>
      wrapper
        .find('.line-placeholder-item')
        .at(index)
        .props();
    expect(getTargetIndexProps(wrapper, 0)['orientation']).toBe('vertical');
    expect(getTargetIndexProps(wrapper, 0)['style'].left).toBe('1px');
    expect(getTargetIndexProps(wrapper, 0)['style'].top).toBe('1px');
    expect(getTargetIndexProps(wrapper, 0)['style'].width).toBe('0.5px');

    expect(getTargetIndexProps(wrapper, 1)['orientation']).toBe('vertical');
    expect(getTargetIndexProps(wrapper, 1)['style'].left).toBe('0px');
    expect(getTargetIndexProps(wrapper, 1)['style'].top).toBe('0px');
    expect(getTargetIndexProps(wrapper, 1)['style'].width).toBe('0.5px');

    expect(getTargetIndexProps(wrapper, 2)['orientation']).toBe('vertical');
    expect(getTargetIndexProps(wrapper, 2)['style'].left).toBe('0px');
    expect(getTargetIndexProps(wrapper, 2)['style'].top).toBe('0.5px');
    expect(getTargetIndexProps(wrapper, 2)['style'].width).toBe('1px');
  });

  it('renderHorizontal should work well', () => {
    const wrapper = mount(
      <SteppedLine x0={1} y0={1} x1={0} y1={0} orientation="horizonal" />,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.line-placeholder-item').length).toBe(3);
    const getTargetIndexProps = (wrapper, index) =>
      wrapper
        .find('.line-placeholder-item')
        .at(index)
        .props();
    expect(getTargetIndexProps(wrapper, 0)['orientation']).toBe('horizonal');
    expect(getTargetIndexProps(wrapper, 0)['style'].left).toBe('1px');
    expect(getTargetIndexProps(wrapper, 0)['style'].top).toBe('1px');
    expect(getTargetIndexProps(wrapper, 0)['style'].width).toBe('0.5px');

    expect(getTargetIndexProps(wrapper, 1)['orientation']).toBe('horizonal');
    expect(getTargetIndexProps(wrapper, 1)['style'].left).toBe('0px');
    expect(getTargetIndexProps(wrapper, 1)['style'].top).toBe('0px');
    expect(getTargetIndexProps(wrapper, 1)['style'].width).toBe('0.5px');

    expect(getTargetIndexProps(wrapper, 2)['orientation']).toBe('horizonal');
    expect(getTargetIndexProps(wrapper, 2)['style'].left).toBe('0.5px');
    expect(getTargetIndexProps(wrapper, 2)['style'].top).toBe('-1px');
    expect(getTargetIndexProps(wrapper, 2)['style'].width).toBe('2px');
  });
});
