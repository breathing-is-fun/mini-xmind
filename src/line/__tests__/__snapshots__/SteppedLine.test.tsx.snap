// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SteppedLine renderHorizontal should work well 1`] = `
<SteppedLine
  className=""
  orientation="horizonal"
  style={Object {}}
  x0={1}
  x1={0}
  y0={1}
  y1={0}
>
  <div
    className="stepped-line"
  >
    <Line
      className=""
      orientation="horizonal"
      style={Object {}}
      x0={1}
      x1={0.5}
      y0={1}
      y1={1}
    >
      <div
        className="line-placeholder"
      >
        <div
          className="line-placeholder-item"
          orientation="horizonal"
          style={
            Object {
              "borderTopColor": "#ddd",
              "borderTopStyle": "solid",
              "borderTopWidth": 1,
              "left": "1px",
              "position": "absolute",
              "top": "1px",
              "transform": "rotate(180deg)",
              "transformOrigin": "0 0",
              "width": "0.5px",
              "zIndex": "1",
            }
          }
        />
      </div>
    </Line>
    <Line
      className=""
      orientation="horizonal"
      style={Object {}}
      x0={0}
      x1={0.5}
      y0={0}
      y1={0}
    >
      <div
        className="line-placeholder"
      >
        <div
          className="line-placeholder-item"
          orientation="horizonal"
          style={
            Object {
              "borderTopColor": "#ddd",
              "borderTopStyle": "solid",
              "borderTopWidth": 1,
              "left": "0px",
              "position": "absolute",
              "top": "0px",
              "transform": "rotate(0deg)",
              "transformOrigin": "0 0",
              "width": "0.5px",
              "zIndex": "1",
            }
          }
        />
      </div>
    </Line>
    <Line
      className=""
      orientation="horizonal"
      style={Object {}}
      x0={0.5}
      x1={0.5}
      y0={-1}
      y1={1}
    >
      <div
        className="line-placeholder"
      >
        <div
          className="line-placeholder-item"
          orientation="horizonal"
          style={
            Object {
              "borderTopColor": "#ddd",
              "borderTopStyle": "solid",
              "borderTopWidth": 1,
              "left": "0.5px",
              "position": "absolute",
              "top": "-1px",
              "transform": "rotate(90deg)",
              "transformOrigin": "0 0",
              "width": "2px",
              "zIndex": "1",
            }
          }
        />
      </div>
    </Line>
  </div>
</SteppedLine>
`;

exports[`SteppedLine renderVertical should work well 1`] = `
<SteppedLine
  className=""
  orientation="vertical"
  style={Object {}}
  x0={1}
  x1={0}
  y0={1}
  y1={0}
>
  <div
    className="stepped-lineto"
  >
    <Line
      className=""
      orientation="vertical"
      style={Object {}}
      x0={1}
      x1={1}
      y0={1}
      y1={0.5}
    >
      <div
        className="line-placeholder"
      >
        <div
          className="line-placeholder-item"
          orientation="vertical"
          style={
            Object {
              "borderTopColor": "#ddd",
              "borderTopStyle": "solid",
              "borderTopWidth": 1,
              "left": "1px",
              "position": "absolute",
              "top": "1px",
              "transform": "rotate(-90deg)",
              "transformOrigin": "0 0",
              "width": "0.5px",
              "zIndex": "1",
            }
          }
        />
      </div>
    </Line>
    <Line
      className=""
      orientation="vertical"
      style={Object {}}
      x0={0}
      x1={0}
      y0={0}
      y1={0.5}
    >
      <div
        className="line-placeholder"
      >
        <div
          className="line-placeholder-item"
          orientation="vertical"
          style={
            Object {
              "borderTopColor": "#ddd",
              "borderTopStyle": "solid",
              "borderTopWidth": 1,
              "left": "0px",
              "position": "absolute",
              "top": "0px",
              "transform": "rotate(90deg)",
              "transformOrigin": "0 0",
              "width": "0.5px",
              "zIndex": "1",
            }
          }
        />
      </div>
    </Line>
    <Line
      className=""
      orientation="vertical"
      style={Object {}}
      x0={0}
      x1={1}
      y0={0.5}
      y1={0.5}
    >
      <div
        className="line-placeholder"
      >
        <div
          className="line-placeholder-item"
          orientation="vertical"
          style={
            Object {
              "borderTopColor": "#ddd",
              "borderTopStyle": "solid",
              "borderTopWidth": 1,
              "left": "0px",
              "position": "absolute",
              "top": "0.5px",
              "transform": "rotate(0deg)",
              "transformOrigin": "0 0",
              "width": "1px",
              "zIndex": "1",
            }
          }
        />
      </div>
    </Line>
  </div>
</SteppedLine>
`;
